File: api/src/modules/content/trail.model.ts
import mongoose, { Schema, Document } from 'mongoose';

export interface ITrail extends Document {
  title: string;
  slug: string;
  description: string;
  thumbnail_url?: string;
  character_id: mongoose.Types.ObjectId;
  
  order: number;
  is_core: boolean;
  denomination?: string;
  unlock_level: number;
  
  total_units: number;
  total_lessons: number;
  estimated_hours: number;
  
  is_published: boolean;
  is_premium: boolean;
  
  company_id?: mongoose.Types.ObjectId;
  
  createdAt: Date;
  updatedAt: Date;
}

const TrailSchema = new Schema<ITrail>(
  {
    title: { type: String, required: true },
    slug: { type: String, required: true, unique: true },
    description: { type: String, required: true },
    thumbnail_url: { type: String },
    character_id: { type: Schema.Types.ObjectId, ref: 'Character' },
    
    order: { type: Number, required: true },
    is_core: { type: Boolean, default: true },
    denomination: { type: String },
    unlock_level: { type: Number, default: 1 },
    
    total_units: { type: Number, default: 0 },
    total_lessons: { type: Number, default: 0 },
    estimated_hours: { type: Number, default: 0 },
    
    is_published: { type: Boolean, default: false },
    is_premium: { type: Boolean, default: false },
    
    company_id: { type: Schema.Types.ObjectId, ref: 'Company' },
  },
  { timestamps: true }
);

TrailSchema.index({ slug: 1 });
TrailSchema.index({ company_id: 1 });
TrailSchema.index({ is_published: 1, order: 1 });

export const TrailModel = mongoose.model<ITrail>('Trail', TrailSchema);

File: app/lib/shared/models/trail_model.dart
/// Model de trilha retornado pela API (metadados do catálogo).
/// Diferente do TrailModel de lesson_model.dart que contém units/lessons embarcados.
class ApiTrailModel {
  final String id;
  final String title;
  final String slug;
  final String description;
  final String? thumbnailUrl;
  final CharacterRef? character;
  final int order;
  final bool isCore;
  final String? denomination;
  final int unlockLevel;
  final double estimatedHours;
  final bool isPremium;
  final bool isPublished;
  final int totalUnits;
  final int totalLessons;

  const ApiTrailModel({
    required this.id,
    required this.title,
    required this.slug,
    required this.description,
    this.thumbnailUrl,
    this.character,
    required this.order,
    required this.isCore,
    this.denomination,
    required this.unlockLevel,
    required this.estimatedHours,
    required this.isPremium,
    required this.isPublished,
    required this.totalUnits,
    required this.totalLessons,
  });

  factory ApiTrailModel.fromJson(Map<String, dynamic> json) => ApiTrailModel(
    id:             (json['_id'] ?? json['id'] ?? '') as String,
    title:          (json['title'] ?? '') as String,
    slug:           (json['slug'] ?? '') as String,
    description:    (json['description'] ?? '') as String,
    thumbnailUrl:   json['thumbnail_url'] as String?,
    character:      json['character_id'] != null && json['character_id'] is Map
                      ? CharacterRef.fromJson(json['character_id'] as Map<String, dynamic>)
                      : null,
    order:          ((json['order'] ?? 0) as num).toInt(),
    isCore:         (json['is_core'] ?? true) as bool,
    denomination:   json['denomination'] as String?,
    unlockLevel:    ((json['unlock_level'] ?? 1) as num).toInt(),
    estimatedHours: ((json['estimated_hours'] ?? 1) as num).toDouble(),
    isPremium:      (json['is_premium'] ?? false) as bool,
    isPublished:    (json['is_published'] ?? false) as bool,
    totalUnits:     ((json['total_units'] ?? 0) as num).toInt(),
    totalLessons:   ((json['total_lessons'] ?? 0) as num).toInt(),
  );
}

class CharacterRef {
  final String id;
  final String name;
  final String? spriteUrl;
  final String? colorHex;

  const CharacterRef({
    required this.id,
    required this.name,
    this.spriteUrl,
    this.colorHex,
  });

  factory CharacterRef.fromJson(Map<String, dynamic> json) => CharacterRef(
    id:        (json['_id'] ?? json['id'] ?? '') as String,
    name:      (json['name'] ?? '') as String,
    spriteUrl: json['sprite_url'] as String?,
    colorHex:  json['color_hex'] as String?,
  );
}
